<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Lesson 6: Data visualisation with ggplot</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">WolfWorks: An Introduction to R</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Setup
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./S00_background.html">Course Outline</a>
    </li>
    <li>
      <a href="./S00_setup.html">Course Setup</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Lessons
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="./S01_R_and_RStudio.html">R and RStudio</a>
    </li>
    <li>
      <a href="./S02_Objects_in_R.html">Objects in R</a>
    </li>
    <li>
      <a href="./S03_Data_structures_in_R.html">Data structures in R</a>
    </li>
    <li>
      <a href="./S04_Starting_with_data.html">Starting with data</a>
    </li>
    <li>
      <a href="./S05_Intro_tidyverse.html">Intro to the tidyverse</a>
    </li>
    <li>
      <a href="./S06_Data_visualisation_ggplot.html">Data visualisation with ggplot2</a>
    </li>
  </ul>
</li>
<li>
  <a href="./contact.html">Contact</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Charl-Hutchings/RIntro">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Lesson 6: Data visualisation with ggplot</h1>

</div>


<style type="text/css">

body, td {
   font-size: 18px;
}
code.r{
  font-size: 14px;
}
pre {
  font-size: 14px
}
</style>
<p><em>This is the course handbook for WolfWorks: An introduction to R.</em></p>
<hr />
<p>Objectives:</p>
<ol style="list-style-type: decimal">
<li>Understand how to create a set of mappings from variables in the data to the <code>aesthetics</code> of a ggplot</li>
<li>Be aware of different geometric objects that we can add to a plot</li>
<li>Create different layers of a ggplot by using the <code>+</code> operator</li>
<li>Change the <code>theme</code> of a ggplot</li>
<li>Use faceting to display different groups of data on different plots - <code>facet_grid()</code> vs. <code>facet_wrap()</code></li>
<li>Know how to export plots using <code>ggsave()</code></li>
</ol>
<hr />
<div id="an-introduction-to-ggplot2" class="section level2">
<h2>An introduction to <code>ggplot2</code></h2>
<p>The final part of this workshop will introduce you to key tools required for plotting anda data visualisation. Why is data visualisation important?</p>
<ul>
<li>Contributes to decision-making during analysis e.g., plotting residuals can indicate whether a statistical test is appropriate or not</li>
<li>Allows the user to understand their data better and be confident in their processing and analysis</li>
<li>Can help to identify when something has gone wrong i.e., if a plot does not look as expected</li>
<li>Ultimately, data is published in a visual form and figures are one of the first things that a researcher will look at in a paper</li>
</ul>
<p>The <code>ggplot2</code> package makes it simple to create publication-quality figures from data stored in a <code>data.frame</code> or <code>tibble</code>. As a note, the <code>2</code> in <code>ggplot2</code> simply refers to the fact that this is the most up-to-date version of the package. <br></p>
<p>We will cover the main points needed to use <code>ggplot2</code>, but for additional information see the <a href="https://carpentries-incubator.github.io/bioc-intro/40-visualization.html">Data Visualisation Cheat Sheet</a>. <br> <br></p>
<p><strong>The grammar of graphics in <code>ggplot</code></strong></p>
<p>The main function within the <code>ggplot2</code> package is <code>ggplot()</code> - this is what we will use to create our visualisations. The theoretical foundation that supports the use of <code>ggplot2</code> is the <em>Grammar of Graphics</em> (what the gg stands for). This is the idea that you can build every graph from the same three components:</p>
<ol style="list-style-type: decimal">
<li>A data set - either a <code>data.frame</code> or <code>tibble</code>. Of note, <code>ggplot2</code> functions like data in a <a href="https://www.statology.org/long-vs-wide-data/">long format</a>.</li>
<li>A coordinate system - provided by mapping variables to visual properties of objects in the plot, termed <strong>aesthetics</strong></li>
<li>A <code>geom</code> to specify the type of plot required <br></li>
</ol>
<p>In terms of code, this looks like:</p>
<p><code>ggplot(data = &lt;DATA&gt;, mapping = aes(&lt;MAPPINGS&gt;)) + &lt;GEOM_FUNCTION&gt;</code></p>
<p>Let’s deal with these three components one at a time. <br> <br></p>
<div id="the-data-set" class="section level4">
<h4>1. The data set</h4>
<p>The first argument that we pass to the <code>ggplot()</code> function is <code>data =</code>. We use this to specify our <code>data.frame</code> or <code>tibble</code>. If we are piping data into the <code>ggplot()</code> function, then we do not need to specify the <code>data =</code> argument.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> rna_tbl)</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
<p>If we execute this code by itself, what we see is that a plotting pane has been opened. <br> <br></p>
</div>
<div id="aesthetics-mappings" class="section level4">
<h4>2. Aesthetics mappings</h4>
<p>We now need to specify <strong>aesthetics mappings</strong>. This will tell the <code>ggplot()</code> function how we want to map variables within our data to visual aspects. The aesthetic mappings take properties of our data and use them to determine visual characteristics such as <em>position</em>, <em>colour</em>, <em>size</em>, <em>shape</em>, or <em>transparency</em> The most simple aesthetic mapping is defining which variables to put on our x- and y-axes.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> rna_tbl,</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> expression))</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>We now see that these variables appear on our plot. <br> <br></p>
</div>
<div id="specifying-geometric-shapes-via-geoms" class="section level4">
<h4>3. Specifying geometric shapes via <code>geoms</code></h4>
<p>Finally, we tell <code>ggplot()</code> what type of plot we are trying to create. This requires us to specify what <strong>geometric objects</strong> we want on the plot - <em>lines</em>, <em>points</em>, <em>bars</em>, <em>boxplots</em> etc. We do this by using the <code>+</code> operator to add one of the following <code>geoms</code> to our plot:</p>
<ul>
<li><code>geom_point</code> - for individual points (i.e., scatter plot)</li>
<li><code>geom_line</code> - for lines</li>
<li><code>geom_smooth</code> - for smoothed lines (e.g., simple trends)</li>
<li><code>geom_bar</code> - for bar charts</li>
<li><code>geom_histogram</code> - for binned values</li>
<li><code>geom_boxplot</code> - for boxplots <br></li>
</ul>
<p>Other <code>geoms</code> you may wish to explore:</p>
<ul>
<li><code>geom_text</code> - add text directly to the plot</li>
<li><code>geom_label</code> - add text to directly to the plot (with a rectangle underneath giving a label-look)</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> rna_tbl,</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> expression)) <span class="sc">+</span> </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<p>The plot above is not particularly useful. Firstly, we have time points 2- and 6-days, which do not exist in our data. This is because <code>time</code> is currently a numeric variable, meaning that R treats it as continuous. In reality, we want to plot time as a factor with three categories - 0, 4 and 8-days post-treatment. Secondly, there are so many genes that we cannot really see the distribution of expression values. This plot would be much better suited as a boxplot. Let’s try to correct these two things.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>rna_tbl <span class="sc">%&gt;%</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">time =</span> <span class="fu">factor</span>(time)) <span class="sc">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> expression)) <span class="sc">+</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>Hmm, this still is not really what we want. Our data is so skewed that the distributions are difficult to see. One trick that is often used in such cases is to log the data in our plot.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>rna_tbl <span class="sc">%&gt;%</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">time =</span> <span class="fu">factor</span>(time)) <span class="sc">%&gt;%</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> <span class="fu">log</span>(expression))) <span class="sc">+</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code></pre></div>
<pre><code>## Warning: Removed 507 rows containing non-finite values (`stat_boxplot()`).</code></pre>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-5-1.png" width="672" /> <br> This looks better. We also get a warning tell us that some values were removed. These are the <code>NA</code> expression values in our data.<br />
<br></p>
<p>We can also add multiple <code>geoms</code> to a single <code>ggplot</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>rna_tbl <span class="sc">%&gt;%</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">time =</span> <span class="fu">factor</span>(time)) <span class="sc">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> <span class="fu">log</span>(expression))) <span class="sc">+</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="at">color =</span> <span class="st">&quot;tomato&quot;</span>) <span class="sc">+</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="dv">0</span>) </span></code></pre></div>
<pre><code>## Warning: Removed 507 rows containing non-finite values (`stat_boxplot()`).</code></pre>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-6-1.png" width="672" /> <br> When we plot multiple <code>geom</code> we need to be aware of the order in which we plot them. Whatever we plot first will end up at the back as <code>ggplot</code> and then new <code>geom</code>s are plotted on top. This is important to remember, for example if we want to use <code>geom_text</code> or <code>geom_label</code> to plot text or labels onto our plots. <br> <br></p>
<p><strong>Challenge: Using the Grammar of Graphics</strong><br />
Create the same plot as you produced previously in base-R, but this time using only <code>tidyverse</code> and <code>ggplot2</code> functions. The plot is showing the expression of Ubfd1 at 8-days post-treatment in male and female mice.</p>
<p><img src="images/example_graph_challenge.png" width="100%" style="display: block; margin: auto;" /> <br></p>
<details>
<summary>
Solution
</summary>
<p><br></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>rna_tbl <span class="sc">%&gt;%</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">==</span> <span class="st">&quot;Ubfd1&quot;</span> <span class="sc">&amp;</span> time <span class="sc">==</span> <span class="dv">8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sex =</span> <span class="fu">factor</span>(sex, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;Male&quot;</span>, <span class="st">&quot;Female&quot;</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sex, <span class="at">y =</span> expression, <span class="at">fill =</span> sex)) <span class="sc">+</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;lightgreen&quot;</span>, <span class="st">&quot;orange&quot;</span>)) <span class="sc">+</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Sex&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Expression of Ubfd1&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Expression of Ubfd1 at 8-days post-treatment&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() </span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
</details>
<p><br></p>
<hr />
</div>
</div>
<div id="more-details-about-aesthetics---in-the-plot-vs.-in-the-layers" class="section level2">
<h2>More details about aesthetics - in the plot vs. in the layers</h2>
<p>As we discovered above, aesthetics can be mapped to the plot such that the plot visuals are dependent upon some variable in our data. This can include specifying which variables determine the x and y coordinates (or in some cases only one of these e.g., histograms only require an x variable as the y coordinate automatically represents frequency). In the example above we saw that we can also have data variables determining the colour of our plots.</p>
<p>Other aesthetic variables that can be altered:</p>
<ul>
<li>point shape: <code>shape</code></li>
<li>point size: <code>size</code> `</li>
<li>line type: <code>line_type</code></li>
<li>line width: <code>size</code><br />
</li>
<li>transparency: <code>alpha</code> <br></li>
</ul>
<p>Any aesthetics that are specified in the initial plot layer will be applied to all <code>geoms</code> that are added. For example, <code>ggplot(aes(alpha = 0.2))</code> would make all <code>geoms</code> somewhat transparent. However, aesthetics can also be specified within specific <code>geom</code> layers so that they are only applied to these layers rather than universally. <br></p>
<hr />
</div>
<div id="using-themes-in-ggplot2" class="section level2">
<h2>Using themes in <code>ggplot2</code></h2>
<p>Another aspect of <code>ggplot2</code> is themes - these include all non-data elements that you can control. Examples include text size, plot background colour, the presence or absence of grid lines, legend position etc. As with all of the other elements that we have come across so far, we add theme elements using the <code>+</code> operator. If we don’t specifically add these elements, the plot will simply use all of <code>ggplot2</code>’s defaults, which is what we did above.<br />
<br></p>
<p>Several themes exist in the <code>ggplot2</code> package:</p>
<ul>
<li><code>theme_grey()</code></li>
<li><code>theme_bw()</code></li>
<li><code>theme_light()</code></li>
<li><code>theme_dark()</code></li>
<li><code>theme_minimal()</code></li>
<li><code>theme_classic()</code> <br></li>
</ul>
<p>Let’s try some of these out using the same plot as above:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Adding black and white theme</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>rna_tbl <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">time =</span> <span class="fu">factor</span>(time)) <span class="sc">%&gt;%</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> <span class="fu">log</span>(expression))) <span class="sc">+</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Adding dark theme</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>rna_tbl <span class="sc">%&gt;%</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">time =</span> <span class="fu">factor</span>(time)) <span class="sc">%&gt;%</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> <span class="fu">log</span>(expression))) <span class="sc">+</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_dark</span>()</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-10-1.png" width="672" /> <br></p>
<p>We can also install other themes from other packages. There is even a package that includes <code>ggplot</code> themes based on films - <a href="https://github.com/MatthewBJane/ThemePark"><code>ThemePark</code></a>.</p>
<p><br> ***</p>
</div>
<div id="faceting" class="section level2">
<h2>Faceting</h2>
<p>Another really useful feature that <code>ggplot2</code> gives us is <strong>faceting</strong>. Faceting lets us split a plot into subplots, or <em>facets</em>, based on a factor (categorical variable) within our dataset.</p>
<p>There are two faceting functions:</p>
<ol style="list-style-type: decimal">
<li><code>facet_wrap()</code> = only creates plots for the combinations of variables that have values by “wrapping” these plots into 2d panels</li>
<li><code>facet_grid()</code> = creates a grid of plots for each combination of variables specified, even if these do not have values</li>
</ol>
<p>Let’s get some data to plot. Create a new <code>tibble</code> called <code>plotting_data</code> that contains the median expression value for each gene at each time.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>plotting_data <span class="ot">&lt;-</span> rna_tbl <span class="sc">%&gt;%</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> rna_tbl<span class="sc">$</span>gene[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]) <span class="sc">%&gt;%</span>                      <span class="co"># Take the top 10 gene names only</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(time, gene) <span class="sc">%&gt;%</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">med_exp =</span> <span class="fu">median</span>(expression, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span>      <span class="co"># Get median expression of these 10 genes at each timepoint</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene, time, med_exp)                                  <span class="co"># Select only relevant columns</span></span></code></pre></div>
<pre><code>## `summarise()` has grouped output by &#39;time&#39;. You can override using the `.groups` argument.</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>plotting_data</span></code></pre></div>
<pre><code>## # A tibble: 30 × 3
## # Groups:   time [3]
##    gene     time med_exp
##    &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;
##  1 Apod        0   11258
##  2 Asl         0     481
##  3 Cyp2d22     0    1883
##  4 Exd2        0    2488
##  5 Fcrls       0     248
##  6 Gjc2        0     595
##  7 Gnb4        0    1791
##  8 Klk6        0     742
##  9 Plp1        0   96534
## 10 Slc2a4      0     249
## # ℹ 20 more rows</code></pre>
<p>We could plot the mecian expression of each of these genes over time on a single plot.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>plotting_data <span class="sc">%&gt;%</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> med_exp)) <span class="sc">+</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>()</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-12-1.png" width="672" /> <br></p>
<p>Uh oh - this is not what we wanted. We plotted the data for all genes together rather than on a per-gene basis. To tell ggplot to give us a line per gene we need to include the <code>group = gene</code> argument in our aesthetics.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>plotting_data <span class="sc">%&gt;%</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> med_exp, <span class="at">group =</span> gene)) <span class="sc">+</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>()</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-13-1.png" width="672" /> <br></p>
<p><strong>Question</strong><br />
What argument could we include if we wanted each of the genes to have a different coloured line? Where would this argument be specified? <br></p>
<p>Faceting allows us to create an entire plot per-gene, rather than having all of the genes on a single plot. Let’s look at the <code>facet_wrap</code> function first.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>plotting_data <span class="sc">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> med_exp)) <span class="sc">+</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> gene)</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p><strong>Question</strong><br />
What do you notice about the y-axis of these plots? <br></p>
<p>If we wanted to allow each plot to have free scaling, we can use the <code>scales = "free_y"</code> argument within <code>facet_wrap</code>.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>plotting_data <span class="sc">%&gt;%</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> med_exp)) <span class="sc">+</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> gene, <span class="at">scales =</span> <span class="st">&quot;free_y&quot;</span>)</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-15-1.png" width="672" /> <br> <br></p>
<p>What about if we try to do the same thing but usig <code>facet_grid</code>?</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>plotting_data <span class="sc">%&gt;%</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> med_exp)) <span class="sc">+</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(.<span class="sc">~</span> gene)</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-16-1.png" width="672" /> <br> Right now we only have a one-dimensional grid because be are only faceting on a single variable. If we wanted to see the median gene expression over time separately for males and females, we could <code>group_by</code> sex as well as gene and time. This way we can facet sex against gene and we get an even more informative plot.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>plotting_data_2 <span class="ot">&lt;-</span> rna_tbl <span class="sc">%&gt;%</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene <span class="sc">%in%</span> rna_tbl<span class="sc">$</span>gene[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]) <span class="sc">%&gt;%</span>                      <span class="co"># Take the top 10 gene names only</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(time, gene, sex) <span class="sc">%&gt;%</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">med_exp =</span> <span class="fu">median</span>(expression, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span>      <span class="co"># Get median expression of these 10 genes at each timepoint for each sex</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene, time, med_exp, sex)                               <span class="co"># Select only relevant columns</span></span></code></pre></div>
<pre><code>## `summarise()` has grouped output by &#39;time&#39;, &#39;gene&#39;. You can override using the `.groups` argument.</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>plotting_data_2</span></code></pre></div>
<pre><code>## # A tibble: 60 × 4
## # Groups:   time, gene [30]
##    gene     time med_exp sex   
##    &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt; 
##  1 Apod        0  10802. Female
##  2 Apod        0  11812  Male  
##  3 Asl         0    447  Female
##  4 Asl         0    481  Male  
##  5 Cyp2d22     0   1750. Female
##  6 Cyp2d22     0   2008  Male  
##  7 Exd2        0   2500. Female
##  8 Exd2        0   2488  Male  
##  9 Fcrls       0    264. Female
## 10 Fcrls       0    248  Male  
## # ℹ 50 more rows</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>plotting_data_2 <span class="sc">%&gt;%</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> med_exp)) <span class="sc">+</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(sex <span class="sc">~</span> gene)</span></code></pre></div>
<p><img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-18-1.png" width="672" /> <br> <br></p>
<p><strong>Challenge: Creating complex ggplots</strong><br />
Using all of the skills we have learnt in this lesson, create a plot that shows mean gene expression of each chromosome throughout the course of infection.</p>
<details>
<summary>
Solution
</summary>
<p><br></p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>rna_tbl <span class="sc">%&gt;%</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(chromosome_name, time) <span class="sc">%&gt;%</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">mean_exp =</span> <span class="fu">mean</span>(expression)) <span class="sc">%&gt;%</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> mean_exp)) <span class="sc">+</span> </span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> chromosome_name, <span class="at">scales =</span> <span class="st">&quot;free_y&quot;</span>)</span></code></pre></div>
<pre><code>## `summarise()` has grouped output by &#39;chromosome_name&#39;. You can override using the `.groups` argument.</code></pre>
<img src="S06_Data_visualisation_ggplot_files/figure-html/unnamed-chunk-19-1.png" width="672" />
</details>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
